gtsfm/averaging/rotation/shonan.py

```diff
@@ -30,20 +30,33 @@ class ShonanRotationAveraging(RotationAveragingBase):
     """Performs Shonan rotation averaging."""
 
     def __init__(self) -> None:
+        """
+        Note: `p_min` and `p_max` describe the minimum and maximum relaxation rank.
+        """
         self._p_min = 5
         self._p_max = 30
 
-    def run(self, num_images: int, i2Ri1_dict: Dict[Tuple[int, int], Optional[Rot3]]) -> List[Optional[Rot3]]:
-        """Run the rotation averaging.
+    def run_w_consecutive_ordering(
+        self, num_connected_nodes: int, i2Ri1_dict: Dict[Tuple[int, int], Optional[Rot3]]
+    ) -> List[Optional[Rot3]]:
+        """Run the rotation averaging on a connected graph w/ N keys ordered consecutively [0,...,N-1].
+
+        Note: the assumption that Shonan keys are ordered as [0,...,N-1] goes deeper than
+        ShonanAveraging::NrUnknowns(), which just checks the assumption. Several other functions that construct
+        matrices related to the convergence certificate also use the fact that the graph is connected and
+        keys are 0..N-1. There are about 10 places where nrUnknowns is used to dimension a sparse or dense matrix.
+        Modifying GTSAM would require a major philosophical overhaul, so we perform the re-ordering
+        here in a sort of "wrapper". See: https://github.com/borglab/gtsam/issues/784
```

Comment:
Why is this note useful here? To someone reading the code, it does not make much sense. NrUnknowns() is not used in this function. Maybe we could add a brief note like "GTSAM requires the nodes to be ordered from [0 ... N-1].  See https://github.com/borglab/gtsam/issues/784 for details. "

Reply:
good point, updated.